### PR TITLE
[JEP-237] Disable usage stats in fips mode

### PR DESCRIPTION
**Pull request by jtnord**
UsageStatistics is so close to being FIPS-140 compliant but fails at the
last hurdle.

We can not encrypt with RSA but we can perform key wrap.  The output does
separate the key and data, by creating a new Key for encrypting the data
and that key is encrypted with RSA.  If only the RSA cipher was created with
Cipher.WRAP_MODE (or Cipher.UNWRAP_MODE) instead of CIPHER.ENCRYPT_MODE (or
CIPHER.DECRYPT_MODE) we would be fine, alas it is not and changing this
would mean a change on the backend and to be able to continue to support
both whilst older versions are in the wild.

The goal of [JEP-237](https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc) is not to fix all non compliant issues, so here we just ensure that in FIPS mode the usage statistics are disabled.

Downstream of #8482

#### regular

no-flags

#### stats disabled by system property

stats_disabled

#### FIPS mode

fips

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Testing done

ran the following with and without the FIPS mode flag and validate results are as expected

```
def stats = ExtensionList.lookupSingleton(UsageStatistics.class)
stats.lastAttempt = 0;
println "isDue: " + stats.isDue()
println "isCollected: " + Jenkins.get().isUsageStatisticsCollected()
println "isNoUsageStatistics: " + Jenkins.get().isNoUsageStatistics()
```

round tripped Jenkins config in both modes. (in FIPS we clear (null) the option)

### Proposed changelog entries

- Disable anonymous usage statistics when run in FIPS mode.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8483"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/security/FIPS140.java
Comment (jtnord) [inline core/src/main/java/jenkins/security/FIPS140.java:1]: ignore this file for this review - please see #8482

**Review by NotMyFault — APPROVED** [review]

**Review by daniel-beck — COMMENTED** [review]
There are two flags that override the effect of the usage stats option: FIPS (new here) and `DISABLED` (pre-existing). This PR introduces a significant behavior difference between the two (e.g. how it shows up on the UI). This seems unnecessary and doesn't help long-term maintainability. Can this be changed so they're treated similarly?

Note also that JEP-214 telemetry is coupled to the usage statistics submission, so if usage stats are forcibly disabled with `FIPS140#useCompliantAlgorithms` enabled, you'll also no longer get telemetry from those instances.

**Comment by jtnord** [comment]
@daniel-beck can you confirm are you asking that the other system property ends up in the same (or similar UX)?

This seems a little at odds with "you'll also no longer get telemetry from those instances" which seemed like a request to decouple the FIPS disabling of stats from the submission of Telemetry, so just want to clarify before making changes.

**Comment by daniel-beck** [comment]
> can you confirm are you asking that the other system property ends up in the same (or similar UX)?

I think that would be nice. Right now these flags seem to be presented very differently (e.g. no custom UI indicating it's disabled for `DISABLED`, only FIPS) even though they effectively do the same thing to both usage stats and JEP-214 telemetry. Plus changes like the one to `#isUsageStatisticsCollected` to only account for FIPS, but it is separate from `DISABLED`, are awkward IMO.

> This seems a little at odds with "you'll also no longer get telemetry from those instances" which seemed like a request to decouple the FIPS disabling of stats from the submission of Telemetry, so just want to clarify before making changes.

It's fine for me if that's what the FIPS flag ends up doing, this one is more of a heads up -- no JEP-214 telemetry as implemented, despite it not being needed for compliance with FIPS (AFAICT).

FWIW I can imagine an implementation of the requirement where everything works as before, except https://github.com/jenkinsci/jenkins/blob/c5fbbbe88dc280d0a0876614c0f9aa6fba2ea2af/core/src/main/java/hudson/model/UsageStatistics.java#L104 is changed to never be `true` with FIPS. That should retain JEP-214 telemetry while not submitting usage stats. (Obviously, then the flags work differently, but at least it's deliberately different behavior to retain more functionality even with FIPS enabled, rather than just incidental differences in their presentation).

**Comment by jtnord** [comment]
> FWIW I can imagine an implementation of the requirement where everything works as before, except
> 
> https://github.com/jenkinsci/jenkins/blob/c5fbbbe88dc280d0a0876614c0f9aa6fba2ea2af/core/src/main/java/hudson/model/UsageStatistics.java#L104
> 
> is changed to never be `true` with FIPS. That should retain [JEP-214](https://github.com/jenkinsci/jep/tree/master/jep/214) telemetry while not submitting usage stats. (Obviously, then the flags work differently, but at least it's deliberately different behavior to retain more functionality even with FIPS enabled, rather than just incidental differences in their presentation).

ironically that is where I started, but figured I would get some pushback as `FIPS` is more of a "feature" than an individual control flag, so I updated the UI (although I did miss this controlling telemetry at the same time - so thanks for the headsup there)  :)

File: core/src/main/java/hudson/model/UsageStatistics.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UsageStatistics.java:107]: diff from pervious commits is FIPS mode only disabled `UsageStatistics` not Telemetry (which was unaffected)

File: core/src/main/java/hudson/model/UsageStatistics.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UsageStatistics.java:218]: tidy up.  (given this code was updated in previous commits it had been changed) this want's a `Boolean` not a `boolean` so avoids autoboxing and an IDE warning.

File: core/src/main/java/jenkins/security/FIPS140.java
Comment (jtnord) [inline core/src/main/java/jenkins/security/FIPS140.java:1]: please review https://github.com/jenkinsci/jenkins/pull/8482 not here.

File: core/src/main/resources/hudson/model/UsageStatistics/global.properties
Comment (jtnord) [inline core/src/main/resources/hudson/model/UsageStatistics/global.properties:26]: we only send Telemetry not `UsageStatistics` so the title adapts based on this.

File: core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly
Comment (jtnord) [inline core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly:10]: remove the General usage statistics when in FIPS mode as it is not applicable

File: core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly
Comment (jtnord) [inline core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly:31]: adapt to the above section not being present in FIPS mode.

**Comment by jtnord** [comment]
> > can you confirm are you asking that the other system property ends up in the same (or similar UX)?
> 
> I think that would be nice. Right now these flags seem to be presented very differently

Feedback should be addressed, see updated screenshots in the description.

File: core/src/main/java/hudson/model/UsageStatistics.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UsageStatistics.java:220]: persist the disabled nature so that any users of the API get the correct state.

File: core/src/main/java/jenkins/security/FIPS140.java
Comment (jtnord) [inline core/src/main/java/jenkins/security/FIPS140.java:1]: ignore this file for this review - please see https://github.com/jenkinsci/jenkins/pull/8482

File: core/src/main/java/hudson/model/UsageStatistics.java
Comment (daniel-beck) [reply core/src/main/java/hudson/model/UsageStatistics.java:220]: Unsure about this change.

By default, usage stats submission is enabled (when the field is still `null`), so any callers of the API that ignore the `DISABLED` flag will get the "correct" value only once the global configuration has been saved.

Right now, it's always wrong to rely on the getter alone (depending on use case). This change may tempt callers into ignoring the edge case of "global config has not yet been submitted".

Comment (jtnord) [reply core/src/main/java/hudson/model/UsageStatistics.java:220]: > By default, usage stats submission is enabled (when the field is still null), so any callers of the API that ignore the DISABLED flag will get the "correct" value only once the global configuration has been saved.

this is true today, and remains so after this change. 

However if the user previously enabled this usage stats, saved the config and then restarted Jenkins with `-Dhudson.model.UsageStatistics.disabled=true` they (as of this change) have no way to change the value in the UI, so it would forever be persisted as `FALSE`.

in no way does this change that any caller of `Jenkins.isUsageStatsCollected` can get invalid info as the value is `null` until a config change.

Comment (daniel-beck) [reply core/src/main/java/hudson/model/UsageStatistics.java:220]: > this is true today, and remains so after this change.

Right, but it's now an edge case more likely to be ignored / callers unaware of.

Don't feel strongly, just seems like something increasing the likelihood of being used wrong (which admittedly was quite substantial even before).

File: core/src/main/resources/hudson/model/UsageStatistics/global.properties
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/UsageStatistics/global.properties:28]:
```suggestion
  The option to send anonymous usage statistics, crash reports and telemetry to the Jenkins project is disabled by a <a href="https://www.jenkins.io/redirect/usagestatistics-disabled">Java system property</a>.
```

We recommend redirect URLs from Jenkins, as they remain stable in the face of docs reorg (needs corresponding https://github.com/jenkins-infra/jenkins.io PR).

File: core/src/main/resources/hudson/model/UsageStatistics/global.groovy
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/UsageStatistics/global.groovy:15]: It might be useful to be explicit here, that general usage statistics are not sent, e.g. in a `f:description`.

**Review by daniel-beck — APPROVED** [review]
Some nice to have nits left, otherwise looks good. Thanks for making the situation better for `DISABLED`!

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
